src/compiler/checker.ts

```diff
@@ -12901,13 +12901,18 @@ namespace ts {
             return result & ObjectFlags.PropagatingFlags;
         }
 
+        function containsMarkerType(typeArguments: readonly Type[] | undefined) {
+            return some(typeArguments, t => t === markerSuperType || t === markerSubType || t === markerOtherType);
+        }
+
         function createTypeReference(target: GenericType, typeArguments: readonly Type[] | undefined): TypeReference {
             const id = getTypeListId(typeArguments);
             let type = target.instantiations.get(id);
             if (!type) {
                 type = createObjectType(ObjectFlags.Reference, target.symbol) as TypeReference;
                 target.instantiations.set(id, type);
-                type.objectFlags |= typeArguments ? getPropagatingFlagsOfTypes(typeArguments, /*excludeKinds*/ 0) : 0;
+                type.objectFlags |= (typeArguments ? getPropagatingFlagsOfTypes(typeArguments, /*excludeKinds*/ 0) : 0) |
+                    (containsMarkerType(typeArguments) ? ObjectFlags.MarkerType : 0);

```

Comment:
Why not just add `MarkerType` to `PropagatingFlags`?

Reply:
I agree - it was a flag on the type/list before so we could avoid doing any kind of deep checks for markers and catch pretty much the exact type we initially make. It was a known shortcoming that we missed markers that flowed into other versions of the reference we're checking the variance of (known to me, at least - I thought it was an intentional performance tradeoff). This now catches if a `Whatever<?, T>` has a `Whatever<T, ?>` dependency (and forces structural comparison, rather than assuming covariance), but still misses a `Whatever<{x: ?}, T>` dependency. Shouldn't we go all the way and catch if the marker appears in the type argument list in any way?

Reply:
Ideally we'd identify any occurrence of a marker type, but it would require eager exploration which we know doesn't end well for many classes of types. Our variance measurement is effectively a balance between correctness and performance, and here we're moving a little further towards correctness. I know we've tried doing _only_ structural comparisons when computing variance, but it isn't feasible. It just reveals the same problems with (almost) infinite types that we're trying to solve by doing variance measurement in the first place.

Reply:
We could add `MarkerType` to `PropagatingFlags`, but it wouldn't save us much and would require us to also include `TypeFlags.TypeParameter` in `TypeFlags.ObjectFlagsType`, which would mean an extra property on every instance that we otherwise don't need.